mitiq/executor/executor.py

```diff
@@ -149,6 +151,29 @@ def evaluate(
                 "Expected observable to be hermitian. Continue with caution."
             )
 
+        # Check executor and observable compatability with type hinting
+        # If FloatLike is specified as a return and observable is used
+        if self._executor_return_type in FloatLike and observable is not None:
+            if self._executor_return_type is not None:
+                raise ValueError(
+                    "When using a float like result, measurements should be "
+                    "included manually and an observable should not be "
+                    "used."
+                )
+        elif observable is None:
+            # Type hinted as DensityMatrixLik but no observable is set
+            if self._executor_return_type in DensityMatrixLike:
+                raise ValueError(
+                    "When using a density matrix like result, an observable "
+                    "is required."
+                )
+            # Type hinted as MeasurementResulteLike but no observable is set
+            elif self._executor_return_type in MeasurementResultLike:
+                raise ValueError(
+                    "When using a measurement, or bitstring, like result, an "
+                    "observable is required."
+                )
```

Comment:
These checks make it impossible to use an `Executor` to run and return density matrices and counts. I don't think we want to limit users in that way, do we?

Reply:
It does make it impossible to use an `Executor` to run and return density matrices and counts if an observable is not provided. This limitation actually comes from parsing the results where `observable` is used by density matrices `_expectation_from_density_matrix()` and counts `_expectation_from_measurements()`.